plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTable.java

```diff
@@ -0,0 +1,137 @@
+import org.testng.annotations.Test;
+
+import java.time.Instant;
+
+import static io.trino.plugin.iceberg.IcebergQueryRunner.createIcebergQueryRunner;
+import static java.lang.String.format;
+
+public class TestIcebergReadVersionedTable
+        extends AbstractTestQueryFramework
+{
+    private long v1Unixtime;
+    private long v1SnapshotId;
+    private long v2Unixtime;
+    private long v2SnapshotId;
+    private long incorrectSnapshotId;
+
+    @Override
+    protected DistributedQueryRunner createQueryRunner()
+            throws Exception
+    {
+        return createIcebergQueryRunner();
+    }
+
+    @BeforeClass
+    public void setUp() throws InterruptedException
+    {
+        assertQuerySucceeds("CREATE TABLE test_iceberg_read_versioned_table (_string VARCHAR, _integer INTEGER)");
+        assertQuerySucceeds("INSERT INTO test_iceberg_read_versioned_table VALUES ('a', 1)");
+        v1Unixtime = Instant.now().getEpochSecond() + 1; // go to the next second in case the operation finishes in the same second
+        v1SnapshotId = getLatestSnapshotId("test_iceberg_read_versioned_table");
+        Thread.sleep(2000);
+        assertQuerySucceeds("INSERT INTO test_iceberg_read_versioned_table VALUES ('b', 2)");
+        v2Unixtime = Instant.now().getEpochSecond() + 1;
+        v2SnapshotId = getLatestSnapshotId("test_iceberg_read_versioned_table");
+        incorrectSnapshotId = v2SnapshotId + 1;
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        assertQuerySucceeds("DROP TABLE IF EXISTS test_iceberg_read_versioned_table");
+    }
+
+    @Test
+    public void testSelectTableWithEndSnapshotId()
+    {
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR VERSION AS OF " + v1SnapshotId, "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR VERSION AS OF " + v2SnapshotId, "VALUES ('a', 1), ('b', 2)");
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR VERSION AS OF " + incorrectSnapshotId, "Iceberg target ID does not exists: " + incorrectSnapshotId);
+    }
+
+    @Test
+    public void testSelectTableWithEndShortTimestamp()
+    {
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(1, 2),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "tpch.\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v1Unixtime, 2), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v2Unixtime, 2), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestamp()
+    {
+        assertQueryFails("SELECT * FROM test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(1, 8),
```

Comment:
inline `timestamp(1, 8)`

just use some fixed timestamp that's before today, as a SQL literal value

Reply:
Sure, I have used `TIMESTAMP '1970-01-01 00:00:00.001000000 Z'` now